### PR TITLE
Enhancement of mobile view and floating windows

### PR DESCRIPTION
**Pull request by ThomasOliverKimble**
**Description**
Enhancement of the floating window class: _(without external library)_
* Scale in 8 directions with boundaries
* Drag with boundaries
* Resize and reposition when out of bounds during screen resize
* Functional with touch screens

Update toolbar for mobile viewing:
* Scale toolbar if `minWidth` is larger than screen
* Add automatic full-screen on landscape orientation
* Add animation to full-screen button to encourage user to use on mobile
* Scale panes and panels if too large for screen

Can be tested on any of the simulations [here](https://testing.webots.cloud/simulation). 
**Using your phone** is encouraged for testing as some functions do not work on desktop mobile simulators.

**Related Issues**
This pull-request needs to be merged along with this `webots.cloud` PR: ([14](https://github.com/cyberbotics/webots-cloud/pull/14))

**Tasks**
Add the list of tasks of this PR.
  - [x] Enhance floating windows
  - [x] Update toolbar for mobile view

### CODE REVIEW

File: resources/web/wwi/Toolbar.js [hunk 1]
```diff
@@ -76,6 +83,22 @@ export default class Toolbar {
...
+      requestFullscreen(this._view);
+    else if (screen.orientation.type === 'portrait-primary' || screen.orientation.type === 'portrait-secondary')
+      exitFullscreen();
+
+    this._scale = this.minWidth > screen.width ? screen.width / this.minWidth : 1;
+    this.toolbar.style.transformOrigin = 'bottom left';
+    this.toolbar.style.transform = 'scale(' + this._scale + ')';
+
+    if (typeof this.robotWindowPane !== 'undefined') {
+      const offset = this._scale == 1 ? 0 : screen.width * (1 - this._scale);
```
Comment (BenjaminDeleze) [inline resources/web/wwi/Toolbar.js:97]:
```suggestion
      const offset = this._scale === 1 ? 0 : screen.width * (1 - this._scale);
```

File: resources/web/wwi/Toolbar.js [hunk 2]
```diff
@@ -345,12 +372,17 @@ export default class Toolbar {
     this.robotWindowList = document.createElement('ul');
     this.robotWindowList.id = 'robot-window-list';
     this.robotWindowPane.appendChild(this.robotWindowList);
+
+    const offset = this._scale == 1 ? 0 : screen.width * (1 - this._scale);
```
Comment (BenjaminDeleze) [inline resources/web/wwi/Toolbar.js:376]:
```suggestion
    const offset = this._scale === 1 ? 0 : screen.width * (1 - this._scale);
```

File: resources/web/wwi/Toolbar.js [hunk 3]
```diff
@@ -424,11 +456,8 @@ export default class Toolbar {
       rw.setPosition(ideOffset + margin + numCol * (margin + robotWindowWidth), margin + numRow * (margin + robotWindowHeight));
       numCol++;
     });
-
-    if (typeof this.ideWindow !== 'undefined' && !this.ideWindow)
-      document.onfullscreenchange = () => { this._fullscreenChangeHeight([this.ideWindow].concat(this.robotWindows)); };
-    else
-      document.onfullscreenchange = () => { this._fullscreenChangeHeight(this.robotWindows); };
+
```
Comment (BenjaminDeleze) [inline resources/web/wwi/Toolbar.js:459]:
```suggestion

```

File: resources/web/wwi/Toolbar.js [hunk 4]
```diff
@@ -500,21 +529,22 @@ export default class Toolbar {
...
-        fw.setSize(fw.getSize()[0], maxHeight);
+  _checkWindowBoundaries() {
+    const resizeObserver = new ResizeObserver(() => {
+      const floatingWindows = document.querySelectorAll('.floating-window');
+      floatingWindows.forEach((fw) => {
+        const maxLeft = this.parentNode.offsetWidth - parseInt(window.getComputedStyle(fw).minWidth);
+        const maxTop = this.parentNode.offsetHeight - parseInt(window.getComputedStyle(fw).minHeight);
+        fw.style.left = (fw.offsetLeft > maxLeft ? maxLeft : fw.offsetLeft) + 'px';
+        fw.style.top = (fw.offsetTop > maxTop ? maxTop : fw.offsetTop) + 'px';
+
```
Comment (BenjaminDeleze) [inline resources/web/wwi/Toolbar.js:540]:
```suggestion

```

**Review by BenjaminDeleze — CHANGES_REQUESTED** [review]
It works well but I have some remarks:

- The animation to encourage the use of fullscreen is nice but I think that once the user has clicked the button once it should stop.
- When testing on mobile in my chrome desktop I got this error on loading: `Uncaught (in promise) TypeError: Failed to execute 'exitFullscreen' on 'Document': Document not active`
- I find a little bit weird to have the signup/login in a sub menu on mobile
- Not a problem of this PR, but if you try to sign up with an already existing address, it bugs (infinite wait) and there is this error in the terminal: `Uncaught (in promise) SyntaxError: Unexpected token < in JSON at position 0`
- I would maybe put the number of pages between the `previous`and `next pages` to have something like: ```Previous   1 2 3   Next page```. What do you think?
- The information panel seems sometime too big for the screen.

The shadows of `OroBOT Simulation on uneven terrain` are completely broken on my phone (not on my computer), is it also the case on yours?

**Comment by ThomasOliverKimble** [comment]
> It works well but I have some remarks:
> 
> * The animation to encourage the use of fullscreen is nice but I think that once the user has clicked the button once it should stop.
> * When testing on mobile in my chrome desktop I got this error on loading: `Uncaught (in promise) TypeError: Failed to execute 'exitFullscreen' on 'Document': Document not active`
> * I find a little bit weird to have the signup/login in a sub menu on mobile
> * Not a problem of this PR, but if you try to sign up with an already existing address, it bugs (infinite wait) and there is this error in the terminal: `Uncaught (in promise) SyntaxError: Unexpected token < in JSON at position 0`
> * I would maybe put the number of pages between the `previous`and `next pages` to have something like: `Previous   1 2 3   Next page`. What do you think?
> * The information panel seems sometime too big for the screen. (on landscape mode no fullscreen).
> * Maybe there is something to do to improve the settings panel (some options goes on 2 lines when the screen is too small)
> 
> The shadows of `OroBOT Simulation on uneven terrain` are completely broken on my phone (not on my computer), is it also the case on yours?

Latest changes:
* Animation to encourage full-screen is removed after being clicked once
* Fixed error on loading: `Uncaught (in promise) TypeError: Failed to execute 'exitFullscreen' on 'Document': Document not active`
* Information panel is scaled if it is larger than the view/screen
* Min width has been added to settings so options are always on one line

Concerning the other changes: I will address them in a seperate PR for `webots.cloud` (signup/login, signup with existing email, previous/next page location, etc.)

Concerning the OroBOT simulation: indeed, the shadows are broken for me on mobile as well, not sure what is happening here... I can modify the `.wbt` file to work though.

**Comment by BenjaminDeleze** [comment]
> Concerning the OroBOT simulation: indeed, the shadows are broken for me on mobile as well, not sure what is happening here... I can modify the `.wbt` file to work though.

I suspect it is maybe a problem of complexity (too high for mobile) or something like that. Anyway, it should not block this PR.

**Review by BenjaminDeleze — CHANGES_REQUESTED** [review]
It works really well!

The only problem I see is: if you are in landscape mode and open a simulation, there is a white space under the simulation until you scroll up

**Review by BenjaminDeleze — APPROVED** [review]
Thank you!

**Pull request merged**